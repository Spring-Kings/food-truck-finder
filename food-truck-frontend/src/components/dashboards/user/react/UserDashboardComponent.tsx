import React, { Component } from "react";

import Router from "next/router";

import Button from "@material-ui/core/Button";
import Grid from "@material-ui/core/Grid";
import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  Card,
  CardContent,
  CircularProgress,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  GridList,
  GridListTile,
  List,
  ListItem,
  Typography,
} from "@material-ui/core";

import TruckRouteMapComponent from "../../../map";
import { SimpleTruck, UserData } from "../../../../redux/user/UserReducer";
import TruckListComponent from "../../TruckListComponent";
import { RouteLocation } from "../../../map/route-map/RouteLocation";
import { DEFAULT_ERR_RESP } from "../../../../api/DefaultResponses";
import { getNearbyTruckLocations, getTruckById } from "../../../../api/Truck";
import TruckLocationMapComponent from "../../../map/truck_location_map/TruckLocationMapComponent";

// Dashboard props
interface UserDashboardProps {
  data: UserData | undefined;
  readonly [x: string]: any;
}

// Dashboard state
interface UserDashboardState {
  inError: string | null;
  nearbyTrucks: RouteLocation[];
  // TODO tighten typing
  viewTruck: any | undefined;
}

// Dashboard component
class UserDashboardComponent extends Component<
  UserDashboardProps,
  UserDashboardState
> {
  constructor(props: UserDashboardProps) {
    super(props);

    // Create state
    this.state = {
      inError: null,
      nearbyTrucks: [],
      viewTruck: undefined
    };

    // Bind methods
    this.viewTruck = this.viewTruck.bind(this);
    this.toOwnerDashboard = this.toOwnerDashboard.bind(this);
  }

  async componentDidMount() {
    // Load
    try {
      if (this.props.data !== undefined) {
        let trucks: RouteLocation[] = await getNearbyTruckLocations(DEFAULT_ERR_RESP);
        this.setState({ inError: null, nearbyTrucks: trucks });
      } else
        await this.props.loadUserFromBackend();
    } catch (err) {
      console.log(err);
      this.setState({ inError: err });
    }
  }

  render() {
    // Ensure the state is OK
    if (this.state.inError)
      return <Typography variant="h1">ERROR: not logged in</Typography>;
    else if (this.props.data === undefined) return <CircularProgress />;

    // If OK, return actual component
    return (
      <React.Fragment>
        {/** Props IDd using: https://material-ui.com/components/grid/ */}
        <GridList
          cols={5}
          style={{
            height: "100vh",
            width: "100%",
          }}
        >
          {/** Side list */}
          <GridListTile cols={1} style={{ height: "100vh" }}>
            {/* Go to owner dashboard */}
            {this.props.data.ownedTrucks ? (
              <Card>
                <Button
                  variant="contained"
                  color="primary"
                  onClick={this.toOwnerDashboard}
                >
                  Owner Dashboard
                </Button>
              </Card>
            ) : null}

            {/* Go to notifications page */}
            <Card>
              <Button
                variant="contained"
                color="primary"
                onClick={this.toNotifications}
              >
                Notifications
              </Button>
            </Card>

            {/* Go to notifications page */}
            <Card>
              <Button
                variant="contained"
                color="primary"
                onClick={this.toSearchTrucks}
              >
                Search Trucks
              </Button>
            </Card>

            {/* Subscribe list */}
            <Accordion>
              <AccordionSummary>Subscribed Trucks</AccordionSummary>
              <AccordionDetails>
                <TruckListComponent
                  trucks={this.props.data.subscribedTrucks}
                  empty={
                    <Card>
                      <Button disabled={true}>No subscriptions</Button>
                    </Card>
                  }
                  handleTruckIcon={<Typography>VIEW</Typography>}
                  handleTruck={this.viewTruck}
                />
              </AccordionDetails>
            </Accordion>
          </GridListTile>

          {/** Where the map would be */}
          <GridListTile cols={4} style={{ height: "100vh" }}>
            <TruckLocationMapComponent
              locations={this.state.nearbyTrucks}
            />
          </GridListTile>
        </GridList>
      </React.Fragment>
    );
  }

  private viewTruck(id: number): void {
    Router.replace(`/truck/${id}`);
  }

  private toOwnerDashboard() {
    Router.replace("/dashboard/owner");
  }

  private toNotifications() {
    Router.replace("/notifications");
  }

  private toSearchTrucks() {
    Router.replace("/search/truck");
  }
}

export default UserDashboardComponent;
