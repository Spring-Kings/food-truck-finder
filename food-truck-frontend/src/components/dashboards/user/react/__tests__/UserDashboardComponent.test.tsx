import React from "react";

jest.mock("../../../../../api/Truck", () => ({
  getNearbyTruckLocations: jest.fn().mockImplementation(() => [mockLOC]),
}));

import UserDashboardComponent from "../UserDashboardComponent";
import user_data from "../../../../../util/test_util/user_data.json";
import {
  RouteLocation,
  RouteLocationState,
} from "../../../../map/route-map/RouteLocation";
import { SNAPSHOT_TEST } from "../../../../../util/test_util/basic_tests";

const mockLOC: RouteLocation = {
  routeLocationId: 0,
  arrivalTime: new Date(),
  exitTime: new Date(),
  stopId: 0,
  coords: { lat: 0, lng: 0 },
  state: RouteLocationState.PERSISTED,
};

SNAPSHOT_TEST(
  "Snapshot test user dashboard",
  () => {
    return (<UserDashboardComponent data={user_data} loadUserFromBackend={() => {}} />);
  }
);