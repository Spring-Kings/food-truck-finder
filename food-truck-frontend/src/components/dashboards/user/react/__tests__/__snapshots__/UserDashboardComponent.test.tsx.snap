// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test user dashboard Matches snapshot 1`] = `"<div class=\\"MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-flex-start\\"><div class=\\"MuiGrid-root MuiGrid-item\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Owner Dashboard</span></button></div></div><ul class=\\"MuiGridList-root\\" style=\\"margin:-2px;height:100vh;width:100%\\"><li class=\\"MuiGridListTile-root\\" style=\\"width:40%;height:50vh;padding:2px\\"><div class=\\"MuiGridListTile-tile\\"><div class=\\"MuiBox-root MuiBox-root-1\\"><h6 class=\\"MuiTypography-root MuiTypography-h6\\">Subscribed Trucks</h6></div><div class=\\"MuiContainer-root MuiContainer-maxWidthLg\\" style=\\"max-height:50vh;overflow:auto\\"><ul class=\\"MuiList-root\\"><li class=\\"MuiListItem-root\\" style=\\"min-width:100%\\"><div class=\\"MuiPaper-root MuiCard-root makeStyles-root-2 MuiPaper-elevation1 MuiPaper-rounded\\"><button class=\\"MuiButtonBase-root MuiCardActionArea-root\\" tabindex=\\"0\\" type=\\"button\\"><div class=\\"MuiCardHeader-root\\"><div class=\\"MuiCardHeader-content\\"><span class=\\"MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock\\"></span></div></div><span class=\\"MuiCardActionArea-focusHighlight\\"></span></button><div class=\\"MuiCardContent-root\\"><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-flex-start\\"><div class=\\"MuiGrid-root MuiGrid-item\\"></div><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-flex-start\\"></div></div></div><div class=\\"MuiCardActions-root MuiCardActions-spacing\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">View Truck</span></button></div></div></li></ul></div></div></li><li class=\\"MuiGridListTile-root\\" style=\\"width:60%;height:50vh;padding:2px\\"><div class=\\"MuiGridListTile-tile\\"><div class=\\"MuiBox-root MuiBox-root-3\\"><h6 class=\\"MuiTypography-root MuiTypography-h6\\">Nearby Trucks</h6></div><div class=\\"MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg\\"><div></div><div>Loading...</div></div></div></li></ul>"`;
